components/pd_client/src/client_v2.rs

```diff
@@ -0,0 +1,1339 @@
+                    panic!(
+                        "fail to request PD {} err {:?}",
+                        "update_service_safe_point", e
+                    );
+                })
+                .await?;
+            PD_REQUEST_HISTOGRAM_VEC
+                .with_label_values(&["update_service_safe_point"])
+                .observe(duration_to_sec(timer.saturating_elapsed()));
+            check_resp_header(resp.get_header())?;
+            Ok(())
+        })
+    }
+
+    fn report_min_resolved_ts(&mut self, store_id: u64, min_resolved_ts: u64) -> PdFuture<()> {
+        let timer = Instant::now();
+
+        let mut req = pdpb::ReportMinResolvedTsRequest::default();
+        req.set_header(self.header());
+        req.set_store_id(store_id);
+        req.set_min_resolved_ts(min_resolved_ts);
+
+        let mut raw_client = self.raw_client.clone();
+        Box::pin(async move {
+            raw_client.wait_for_ready().await?;
+            let resp = raw_client
+                .stub()
+                .report_min_resolved_ts_async_opt(
+                    &req,
+                    raw_client
+                        .call_option()
+                        .timeout(Duration::from_secs(REQUEST_TIMEOUT)),
+                )
+                .unwrap_or_else(|e| {
+                    panic!("fail to request PD {} err {:?}", "min_resolved_ts", e);
+                })
+                .await?;
+            PD_REQUEST_HISTOGRAM_VEC
+                .with_label_values(&["min_resolved_ts"])
+                .observe(duration_to_sec(timer.saturating_elapsed()));
+            check_resp_header(resp.get_header())?;
+            Ok(())
+        })
+    }
+}
+
+async fn reconnect_loop(
+    mut client: CachedRawClient,
+    on_reconnect: CallbackHolder,
+    mut reconnect: mpsc::UnboundedReceiver<u64>,
+) {
+    let interval = Duration::from_secs(REQUEST_TIMEOUT);
+    loop {
+        while !client.channel().wait_for_connected(interval).await {
+            // wait_for_connected will attempt to connect internally. When it fails we
+            // should force reconnect.
+            if let Err(e) = client.reconnect(true, &on_reconnect).await {
+                warn!("failed to reconnect pd"; "err" => ?e);
+            }
+        }
+        let state = ConnectivityState::GRPC_CHANNEL_READY;
+        select! {
+            _ = client.channel().wait_for_state_change(state, interval).fuse() => {
+                let force = client.channel().check_connectivity_state(true) == ConnectivityState::GRPC_CHANNEL_SHUTDOWN;
+                if let Err(e) = client.reconnect(force, &on_reconnect).await {
+                    warn!("failed to reconnect pd"; "err" => ?e);
+                }
+            },
+            v = reconnect.next() => {
+                if v.is_none() {
+                    break;
+                } else if v.unwrap() == client.cache_version() {
+                    if let Err(e) = client.reconnect(true, &on_reconnect).await {
```

Comment:
No need to force reconnect.

Reply:
So you mean the `force=true` condition will never be used in v2? But why is it used in v1 (e.g. force reconnect when tso stream ended)?

Reply:
Because in v1, we don't rely on the auto retry of grpc channel, instead, we control the retry by our own. In v2, we don't do retry any more if the target address is still the same.